### PR TITLE
Add RSpec 4 compatibility

### PR DESCRIPTION
**Pull request by pirj**
You may check a PR that also switches to use RSpec 4 and RSpec-Rails 6: https://github.com/pirj/discourse/pull/1

There are a few reasons it doesn't make sense to [switch right away](https://github.com/pirj/discourse/pull/1/commits/08f058a7ee448b3286d576d4320ac1c7f365103c):
 - RSpec 4 and RSpec-Rails 6 are not yet released, and it's a hassle to [source them from GitHub](https://github.com/pirj/discourse/pull/1/files#diff-d09ea66f8227784ff4393d88a19836f321c915ae10031d16c93d67e6283ab55fR160-R164)
 - RSpec-Rails 6 is [not marked to support RSpec 4](https://github.com/pirj/rspec-rails/commit/6023a2003d5b62b8619f83ae470ef31a6afcba60)
 - ~`rspec-html-matchers` won't work until [this PR](https://github.com/kucaahbe/rspec-html-matchers/pull/77) is merged~ - [fixed in 0.10.0](https://github.com/kucaahbe/rspec-html-matchers/commit/54e99fba653d4b712fd83c87c1bffbe68726f51b)
 - 61 plugins need to be updated ([example PR](https://github.com/discourse/discourse-teambuild/pull/42))

#### Comply to strict predicate matchers

   See:
   - https://github.com/rspec/rspec-expectations/pull/1195
   - https://github.com/rspec/rspec-expectations/pull/1196
   - https://github.com/rspec/rspec-expectations/pull/1277

Basically,
```ruby
expect(OpenStruct.new(foo?: nil)).to_not be_foo
```
won't pass anymore, as strict predicate matchers would only accept `true`/`false`, but not "truthy" or "falsey" values.

#### Use the non-globally exposed RSpec syntax

See https://github.com/rspec/rspec-core/pull/2803

`describe`/`shared_examples` won't work without the explicit `RSpec.`.
`should`/`should_not` were deprecate nearly a decade ago and will be removed.

#### Misc

Discourse specs use defaults, and it helped a lot, there's just `--color` option that had to be removed from `.rspec` file.
https://github.com/rspec/rspec-core/commit/04078317ba6577699d06cf4dccf014254dcde7a6

#### RSpec-Rails 6

No changes turned out to be necessary.

### CODE REVIEW

File: spec/lib/svg_sprite/svg_sprite_spec.rb [hunk 1]
```diff
@@ -254,7 +254,7 @@
       theme.set_field(target: :common, name: SvgSprite.theme_sprite_variable_name, upload_id: upload.id, type: :theme_upload_var)
       theme.save!
 
-      expect(Upload.where(id: upload.id)).to be_exist
+      expect(Upload.where(id: upload.id)).to be_exists
```
Comment (pirj) [inline spec/lib/svg_sprite/svg_sprite_spec.rb:257]: Because `Upload.where(...).exist?` method doesn't exist, but `exists?` does.

Comment (tgxworld) [inline spec/lib/svg_sprite/svg_sprite_spec.rb:257]:
```suggestion
      expect(Upload.exists?(id: upload.id)).to eq(true)
```

File: spec/lib/svg_sprite/svg_sprite_spec.rb [hunk 2]
```diff
@@ -266,7 +266,7 @@
...
```
Comment (tgxworld) [inline spec/lib/svg_sprite/svg_sprite_spec.rb:269]:
```suggestion
      expect(Upload.exists?(id: upload.id)).to eq(true)
```

File: spec/lib/svg_sprite/svg_sprite_spec.rb [hunk 3]
```diff
@@ -280,7 +280,7 @@
       child_theme.set_field(target: :common, name: SvgSprite.theme_sprite_variable_name, upload_id: upload.id, type: :theme_upload_var)
       child_theme.save!
 
-      expect(Upload.where(id: upload.id)).to be_exist
+      expect(Upload.where(id: upload.id)).to be_exists
```
Comment (tgxworld) [inline spec/lib/svg_sprite/svg_sprite_spec.rb:283]:
```suggestion
      expect(Upload.exists?(id: upload.id)).to eq(true)
```

File: spec/models/topic_spec.rb [hunk 4]
```diff
@@ -2351,18 +2351,22 @@ def set_state!(group, user, state)
   describe '#secure_category?' do
     let(:category) { Category.new }
 
+    def read_restricted_category?(topic)
+      topic.read_restricted_category?
+    end
```
Comment (tgxworld) [inline spec/models/topic_spec.rb:2356]:
```suggestion
```

File: spec/models/topic_spec.rb [hunk 5]
```diff
@@ -2351,18 +2351,22 @@ def set_state!(group, user, state)
...
     it "is true if the category is secure" do
       category.stubs(:read_restricted).returns(true)
-      expect(Topic.new(category: category)).to be_read_restricted_category
+      expect(read_restricted_category?(Topic.new(category: category))).to be(true)
     end
 
     it "is false if the category is not secure" do
       category.stubs(:read_restricted).returns(false)
-      expect(Topic.new(category: category)).not_to be_read_restricted_category
+      expect(read_restricted_category?(Topic.new(category: category))).to be(false)
```
Comment (tgxworld) [inline spec/models/topic_spec.rb:2365]:
```suggestion
      expect(Topic.new(category: category).read_restricted_category?).to eq(false)
```

File: spec/models/topic_spec.rb [hunk 6]
```diff
@@ -2351,18 +2351,22 @@ def set_state!(group, user, state)
...
     end
 
-    it "is false if there is no category" do
-      expect(Topic.new(category: nil)).not_to be_read_restricted_category
+    it "is falsey if there is no category" do
+      expect(read_restricted_category?(Topic.new(category: nil))).to be(nil)
```
Comment (tgxworld) [inline spec/models/topic_spec.rb:2369]:
```suggestion
      expect(Topic.new(category: nil).read_restricted_category?).to eq(nil)
```

**Comment by tgxworld** [comment]
Thank you @pirj, some comments but otherwise the PR looks good :+1:

File: spec/models/topic_spec.rb [hunk 7]
```diff
@@ -2351,18 +2351,22 @@ def set_state!(group, user, state)
...
     let(:category) { Category.new }
 
+    def read_restricted_category?(topic)
+      topic.read_restricted_category?
+    end
+
     it "is true if the category is secure" do
       category.stubs(:read_restricted).returns(true)
-      expect(Topic.new(category: category)).to be_read_restricted_category
+      expect(read_restricted_category?(Topic.new(category: category))).to be(true)
```
[the remark this answers is not in the document]
Comment (pirj) [reply spec/models/topic_spec.rb:2360]: I'd stick with `be(true)`, but `eq(true)` is fine, too.

**Comment by pirj** [comment]
Thanks for the quick review, @tgxworld . Do you want me to squash the last commit?

**Comment by pirj** [comment]
Rebased to resolve conflicts, squashed code review fixups.
I dared updating `rspec-html-matchers` to a version that is compatible with RSpec 4 in the last commit, see [their update](https://github.com/kucaahbe/rspec-html-matchers/commit/54e99fba653d4b712fd83c87c1bffbe68726f51b) for details.

**Comment by tgxworld** [comment]
Don't worry about squashing because we'll do a squash and merge once the PR is ready.

**Comment by pirj** [comment]
> the PR is ready

Is there something I can do to help making it ready?

**Review by tgxworld — APPROVED** [review]

**Comment by tgxworld** [comment]
@pirj Thank you for your contribution here :+1:
